### ISSUE TITLE
confusion matrix on test(unseen) data?

### ISSUE THREAD

**Issue by salmankdurrani** [ultralytics/ultralytics#9479]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

I have trained a YOLO-V8 on my custom data and the accuracy is very good but now I want to test it on unseen data and plot a confusion matrix but I am unable to do that. 
I can draw the bonding box and class id information using 

```
new = new_results[0].boxes.data.tolist()
new
```

but I don't know how to plot the confusion matrix on the test folder. please help!

**Comment by glenn-jocher** [comment]
@salmankdurrani hey there! 😊 Plotting a confusion matrix for your YOLOv8 model on unseen data is a great step for evaluating its performance. After you've gathered predictions and true labels, you can use libraries like `matplotlib` and `seaborn` for visualization, along with `sklearn` for computing the confusion matrix. Here's a simple way to do it:

1. First, ensure you have the true labels for your unseen dataset in a comparable format to your predictions.
2. Compute the confusion matrix using `sklearn.metrics.confusion_matrix`.

Assuming `y_true` are your true labels and `y_pred` are your predicted labels (extracted from `new`), you can do the following:

```python
from sklearn.metrics import confusion_matrix
import seaborn as sns
import matplotlib.pyplot as plt

# Example true and predicted labels
y_true = [...]  # Your true labels here
y_pred = [int(item[4]) for item in new]  # Convert your predictions from 'new'

# Compute confusion matrix
conf_matrix = confusion_matrix(y_true, y_pred)

# Plotting
fig, ax = plt.subplots(figsize=(10, 7))
sns.heatmap(conf_matrix, annot=True, fmt='d', ax=ax, cmap="Blues")
plt.ylabel('Actual')
plt.xlabel('Predicted')
plt.title('Confusion Matrix')
plt.show()
```

Make sure to adjust `y_true` and `y_pred` extraction according to your data structure. This example assumes your predictions and true labels are in the simplest form of lists containing class IDs. If your data structure is different, you may need to adjust the extraction logic. Hope this helps! 🌟

**Comment by salmankdurrani** [comment]
Thank you, this is very helpful. But I guess, I didn't put my question correctly!

1. For my YOLO-V8 set, I have a total of 225 frames for the class **"Cut"** in the validation set, but it only shows 114 in the confusion matrix. I wonder where does the rest of the images go? Why is it skipping images in almost every class?

2. Next, In order to solve this, I can just simply run the prediction for each image in the frame, but I do need help with how to take out the `y_true = [...]  # Your true labels here
y_pred = [int(item[4]) for item in new]  # Convert your predictions from 'new'`
form the code 
`
new_image = '/mnt/c/Users/admin/Documents/SalmanData/YOLO_V8_Noise_Study_exp1/images/validate/cut_169.jpeg'
new_results = my_new_model(new_image, conf=0.2)  #Adjust conf threshold
`
![Screenshot 2024-04-02 at 13 26 39](https://github.com/ultralytics/ultralytics/assets/99785647/d7a2992a-817b-4eaa-b1ca-9932c651bfd7)
![Screenshot 2024-04-02 at 13 27 00](https://github.com/ultralytics/ultralytics/assets/99785647/d19935a4-857c-4834-b515-90e1a2055b0d)

**Comment by glenn-jocher** [comment]
@salmankdurrani hey, thanks for reaching out with more details! 😊

1. **Missing Frames in the Confusion Matrix:** The discrepancy you're observing in the number of frames for the **"Cut"** class (or others) in the confusion matrix could be due to frames not being detected with enough confidence (above the threshold). YOLOv8 will only include detections in the confusion matrix if they surpass the `conf` threshold defined in your prediction step. Lower confidence predictions are omitted, which might explain the fewer frames accounted for. It's always a balance in setting the right confidence threshold to capture as many true positives without introducing too many false positives.

2. **Extracting `y_true` and `y_pred`:** To solve your second query, you'll need to manually map your true labels (`y_true`) corresponding to each frame/image in your validation dataset. As for `y_pred`, you're on the right track with how to extract predicted class IDs. If `my_new_model(new_image, conf=0.2)` returns detection results that include class IDs in the fifth position (`item[4]`) of each detection array `new`, your method for obtaining `y_pred` sounds correct. Just ensure you match each prediction with its corresponding ground truth label for accurate comparison.

Here's a snippet focusing on how to get predictions (adjust the confidence threshold if needed):
```python
new_image = '/path/to/your/image.jpeg'
new_results = my_new_model(new_image, conf=0.2)  # Adjust conf threshold
new_predictions = new_results[0].boxes.data.tolist()
y_pred = [int(item[4]) for item in new_predictions]  # Extracting predicted class IDs
```
For `y_true`, you'd need to align this with your dataset's annotation for each image being predicted upon.

Let us know if you need further clarification or assistance!

**Comment by salmankdurrani** [comment]
Thank you so much for the help. 
I was able to plot the confusion matrix. but now I ran into these two new problems. 
1. the confusion matrix is missing background, do the best.pt has the background also included as a class? 
Since the data I have is frames which does only contain the class names but not background. 
so how can I fix this? 

2. can I modify the class name/id in the best.pt? as I don't want to do the trainings again. 
one of the class ID I have given during the training is 'pick/needle visible' which can not be named as image file. so the confusion matrix is unable to match that. 

![Screenshot 2024-04-03 at 5 45 54](https://github.com/ultralytics/ultralytics/assets/99785647/43e488d2-e46f-40b9-ae74-6ba944cfb13f)

**Comment by glenn-jocher** [comment]
@salmankdurrani hey there! 😊 Glad to hear you were able to plot the confusion matrix! Let's quickly address your new questions:

1. **Missing Background Class:** Typically, YOLO models don't explicitly include a "background" class—objects are either detected as specific classes, or not detected at all. If you're looking to incorporate background as a class, you'd usually need to label and train it as such. For frames without detectable objects, they aren't represented in the confusion matrix because they don't contribute to true/false positives/negatives directly. 

   If your use case mandates recognizing a background or "no-object" scenario explicitly, you might consider adding a dedicated "background" class during training, involving significant adjustments and re-training.

2. **Modifying Class Name/ID in `best.pt`:** Directly changing class names or IDs within the `best.pt` model file isn't straightforward as these are learned features based on your training data labels. However, you can map the class names post-prediction for display purposes. For example, if your class ID `2` is "pick/needle visible", you can map it when preparing your labels for the confusion matrix:

```python
class_mapping = {
    2: "pick_needle_visible",  # Modify as per your class ID
    # Add other class mappings here
}
# Replace your `y_pred` class IDs with new names
y_pred_mapped = [class_mapping.get(pred, pred) for pred in y_pred]
```

This way, you can avoid retraining but have more manageable class names for analysis.

Hope this helps, and if you have any more questions, feel free to ask! 🌟

**Comment by salmankdurrani** [comment]
Thank you it is solved!

**Comment by glenn-jocher** [comment]
@salmankdurrani that's fantastic to hear! 🎉 If you have any more questions or need further assistance in the future, feel free to reach out. Happy coding!

**Comment by glenn-jocher** [comment]
@Harikak09 it sounds like there might be a mismatch between the number of predictions and true labels in your dataset, which can cause issues when computing a confusion matrix. Ensure that each image in your test folder has a corresponding ground truth label and that your model predictions align with these labels. You might need to adjust your prediction and label extraction logic to ensure they are consistent. If the issue persists, reviewing your dataset's structure and ensuring consistent labeling can help resolve this. For further guidance, you might find it useful to explore the [Ultralytics guide on model testing](https://docs.ultralytics.com/guides/model-testing/) for more insights.

**Comment by Harikak09** [comment]
Thanks for the comment. But assume that I have images where there is no region of my interest so I won't label or annotate and I mark it as null  for that image. So now my current labels file for that images is empty. But after applying model on the same image, there is a chance that it detects an object even ground truth is not present then _true labels<pred labels_. So in that cases what to do? 

For example I have image of a road with closed manhole in it so my ground truth will be null as my interest is to detect potholes but model detected the manhole as pothole and given some score(probability). In that cases, ground truth labels will be less compared to predicted labels. So now how to do it??

**Comment by glenn-jocher** [comment]
In cases where your ground truth labels are null (indicating no objects of interest) but the model predicts an object, these should be treated as false positives. When calculating your confusion matrix, the absence of ground truth labels should be recorded as a "negative" for your class of interest, and any incorrect predictions should contribute to the false positive count. This approach ensures your model's performance metrics accurately reflect its tendency to incorrectly identify non-target objects. For more details on handling such scenarios, you can explore the [YOLO Common Issues guide](https://docs.ultralytics.com/guides/yolo-common-issues/).
